// ecl/ecl.js
(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
var Base,
  hasProp = {}.hasOwnProperty;

module.exports = Base = (function() {
  function Base(args) {
    var key, value;
    for (key in args) {
      if (!hasProp.call(args, key)) continue;
      value = args[key];
      this[key] = value;
    }
    Object.defineProperty(this, '___runtime', {
      enumerable: false,
      writable: false,
      configurable: false,
      value: (args != null ? args.___runtime : void 0) || {}
    });
  }

  return Base;

})();

},{}],2:[function(require,module,exports){
window.ecl = require('./index');

},{"./index":5}],3:[function(require,module,exports){
var Event,
  extend = require("extends__"),
  hasProp = {}.hasOwnProperty;

module.exports = Event = (function(superClass) {
  extend(Event, superClass);

  function Event(type, callback) {
    var args, date, perf;
    args = (typeof type === 'string' ? {
      type: type
    } : type || {});
    if (typeof callback === 'function') {
      args.callback = callback;
    }
    Event.__super__.constructor.call(this, args);
    if (this.timestamp === true) {
      date = Date.now();
      perf = (typeof performance !== "undefined" && performance !== null ? performance.now() : void 0) || 0;
      this.timestamp = 1000 * date + Math.floor(1000 * (perf - Math.floor(perf)));
    }
  }

  Event.prototype.cancel = function() {
    this.___runtime.cancel = true;
    return this;
  };

  Event.prototype.stop = function() {
    this.___runtime.stop = true;
    return this;
  };

  Event.prototype.abort = function() {
    this.aborted = true;
    return this;
  };

  return Event;

})(require('./base'));

},{"./base":1,"extends__":7}],4:[function(require,module,exports){
var Event, Evented,
  extend = require("extends__"),
  hasProp = {}.hasOwnProperty;

Event = require('./event');

module.exports = Evented = (function(superClass) {
  extend(Evented, superClass);

  function Evented() {
    Evented.__super__.constructor.apply(this, arguments);
    this.listeners = [{}, {}];
  }

  Evented.prototype.addListener = function(type, listener, capture) {
    var listeners, ref, ref1, ref2;
    if (capture == null) {
      capture = false;
    }
    if (typeof type === 'object') {
      ref = type, type = ref.type, listener = ref.listener, capture = ref.capture;
    }
    if (((ref1 = this.events) != null ? ref1[type] : void 0) && typeof listener === 'function') {
      listeners = ((ref2 = this.listeners[capture ? 1 : 0]) != null ? ref2[type] || (ref2[type] = []) : void 0);
      if (-1 === listeners.indexOf(listener)) {
        listeners.push(listener);
      }
    }
    return this;
  };

  Evented.prototype.removeListener = function(type, listener, capture) {
    var idx, listeners, ref, ref1;
    if (capture == null) {
      capture = false;
    }
    if (type instanceof Object) {
      ref = type, type = ref.type, listener = ref.listener, capture = ref.capture;
    }
    if (type && typeof listener === 'function') {
      if (listeners = (ref1 = this.listeners[capture ? 1 : 0]) != null ? ref1[type] : void 0) {
        if (-1 !== (idx = listeners.indexOf(listener))) {
          listeners.splice(idx, 1);
        }
      }
    }
    return this;
  };

  Evented.prototype.dispatchEvent = function(event) {
    var i, len, listener, listeners, phase, ref, ref1, type;
    if (!((event != null ? event.aborted : void 0) || (event != null ? event.canceled : void 0))) {
      if ((type = event != null ? event.type : void 0) && this.events[type]) {
        phase = event.phase;
        if (((3 > phase && phase > 0)) && (listeners = (ref = this.listeners) != null ? ref[2 - phase][type] : void 0)) {
          for (i = 0, len = listeners.length; i < len; i++) {
            listener = listeners[i];
            if (((ref1 = event.___runtime) != null ? ref1.canceled : void 0) || event.aborted) {
              break;
            }
            listener.call(this, event);
          }
        }
      }
    }
    return this;
  };

  Evented.prototype.broadcastEvent = function(event, target) {
    var base, child, i, len, phase, ref, ref1, type;
    if ((type = event != null ? event.type : void 0) && (event.phase || 0) < 3) {
      if (!(event.aborted || event.___runtime.stopped)) {
        (base = event.___runtime).source || (base.source = this);
        phase = (event.phase || (event.phase = 1));
        if (event.target === this) {
          event.phase = 2;
        }
        if (event.phase === 1) {
          this.dispatchEvent(event);
          if (this.children) {
            ref = this.children;
            for (i = 0, len = ref.length; i < len; i++) {
              child = ref[i];
              if (!(event.aborted || event.___runtime.canceled)) {
                child.broadcastEvent(event);
              } else {
                break;
              }
            }
          }
        }
        if (event.target === this) {
          event.phase = 2;
        }
        if (event.phase === 2) {
          this.dispatchEvent(event);
        }
        if (event.___runtime.source === this && event.phase < 4) {
          if ((ref1 = event.callback) != null) {
            if (typeof ref1.call === "function") {
              ref1.call(this, event);
            }
          }
          event.phase = 4;
        }
      }
    }
    return this;
  };

  return Evented;

})(require('./node'));

},{"./event":3,"./node":6,"extends__":7}],5:[function(require,module,exports){
module.exports = {
  Base: require('./base'),
  Node: require('./node'),
  Event: require('./event'),
  Evented: require('./evented')
};

},{"./base":1,"./event":3,"./evented":4,"./node":6}],6:[function(require,module,exports){
var Node,
  extend = require("extends__"),
  hasProp = {}.hasOwnProperty;

module.exports = Node = (function(superClass) {
  extend(Node, superClass);

  function Node() {
    Node.__super__.constructor.apply(this, arguments);
  }

  Node.prototype.appendChild = function(child) {
    var ref;
    if (child !== this) {
      this.children || (this.children = []);
      if (-1 === this.children.indexOf(child)) {
        if ((ref = child.parent) != null) {
          if (typeof ref.removeChild === "function") {
            ref.removeChild(child);
          }
        }
        child.parent = this;
        this.children.push(child);
      }
    }
    return this;
  };

  Node.prototype.removeChild = function(child) {
    var idx, ref;
    if (((ref = this.children) != null ? ref.length : void 0) && -1 !== (idx = this.children.indexOf(child))) {
      delete child.parent;
      this.children.splice(idx, 1);
      if (this.children.length === 0) {
        delete this.children;
      }
    }
    return this;
  };

  return Node;

})(require('./base'));

},{"./base":1,"extends__":7}],7:[function(require,module,exports){
var extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  hasProp = {}.hasOwnProperty;

module.exports = function(ChildClass, ParentClasses) {
  var MixinClass, ParentClass, i, key, len, ref, value;
  if (ParentClasses instanceof Array && ParentClasses.length) {
    ParentClass = (function(superClass) {
      extend(ParentClass, superClass);

      function ParentClass() {
        var MixinClass, i, len;
        ParentClass.__super__.constructor.apply(this, arguments);
        for (i = 0, len = ParentClasses.length; i < len; i++) {
          MixinClass = ParentClasses[i];
          MixinClass.apply(this, arguments);
        }
      }

      return ParentClass;

    })(ParentClasses.shift());
    for (i = 0, len = ParentClasses.length; i < len; i++) {
      MixinClass = ParentClasses[i];
      ref = MixinClass.prototype;
      for (key in ref) {
        if (!hasProp.call(ref, key)) continue;
        value = ref[key];
        if (key !== 'constructor') {
          ParentClass.prototype[key] = value;
        }
      }
    }
  } else {
    ParentClass = ParentClasses;
  }
  return extend(ChildClass, ParentClass);
};

},{}]},{},[2])
